src/hash_map.c

```diff
@@ -0,0 +1,562 @@
+    return RCUTILS_RET_NOT_INITIALIZED; \
+  }
+
+#define HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(arg) \
+  if (NULL == arg) { \
+    RCUTILS_SET_ERROR_MSG("argument is null"); \
+    return RCUTILS_RET_INVALID_ARGUMENT; \
+  }
+
+typedef struct rcutils_hash_map_entry_t
+{
+  size_t hashed_key;
+  void * key;
+  void * value;
+} rcutils_hash_map_entry_t;
+
+typedef struct rcutils_hash_map_impl_t
+{
+  // This is the array of buckets that will store the keypairs
+  rcutils_array_list_t * map;
+  size_t capacity;
+  size_t size;
+  size_t key_size;
+  size_t data_size;
+  rcutils_hash_map_key_hasher_t key_hashing_func;
+  rcutils_hash_map_key_cmp_t key_cmp_func;
+  rcutils_allocator_t allocator;
+} rcutils_hash_map_impl_t;
+
+// djb2 hash function
+size_t rcutils_hash_map_string_hash_func(const void * key_str)
+{
+  const char ** ckey_ptr = (const char **) key_str;
+  const char * ckey_str = *ckey_ptr;
+  size_t hash = 5381;
+
+  while ('\0' != *ckey_str) {
+    int c = *(ckey_str++);
+    hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+  }
+
+  return hash;
+}
+
+int rcutils_hash_map_string_cmp_func(const void * val1, const void * val2)
+{
+  const char ** cval1 = (const char **) val1;
+  const char ** cval2 = (const char **) val2;
+  return strcmp(*cval1, *cval2);
+}
+
+rcutils_hash_map_t
+rcutils_get_zero_initialized_hash_map()
+{
+  static rcutils_hash_map_t zero_initialized_hash_map;
+  zero_initialized_hash_map.impl = NULL;
+  return zero_initialized_hash_map;
+}
+
+// Allocates a new zero initialized hashmap
+static rcutils_ret_t hash_map_allocate_new_map(
+  rcutils_array_list_t ** map, size_t capacity,
+  const rcutils_allocator_t * allocator)
+{
+  *map = allocator->allocate(capacity * sizeof(rcutils_hash_map_impl_t), allocator->state);
+  if (NULL == *map) {
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  // Make sure the list for ever bucket is zero initialized
+  rcutils_array_list_t zero_list = rcutils_get_zero_initialized_array_list();
+  for (size_t i = 0; i < capacity; ++i) {
+    (*map)[i] = zero_list;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+// Deallocates the memory for a map entry
+static void hash_map_deallocate_entry(
+  rcutils_allocator_t * allocator,
+  rcutils_hash_map_entry_t * entry)
+{
+  if (NULL != entry) {
+    if (NULL != entry->key) {
+      allocator->deallocate(entry->key, allocator->state);
+    }
+    if (NULL != entry->value) {
+      allocator->deallocate(entry->value, allocator->state);
+    }
+    allocator->deallocate(entry, allocator->state);
+  }
+}
+
+// Deallocates an existing hashmap
+static rcutils_ret_t hash_map_deallocate_map(
+  rcutils_array_list_t * map, size_t capacity,
+  rcutils_allocator_t * allocator, bool dealloc_map_entries)
+{
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  for (size_t i = 0; i < capacity && RCUTILS_RET_OK == ret; ++i) {
+    // If an array list is allocated for this bucket then clean it up
+    rcutils_array_list_t * bucket = &(map[i]);
+    if (NULL != bucket->impl) {
```

Comment:
@nburek if the `ARRAY_LIST_VALIDATE_ARRAY_LIST()` is turned into a function, use it instead of checking for `impl`.

Reply:
Even though I made the existing macro public by moving it to the header file, I didn't change this to use it as the macro's behavior is to return early, which is not what I wanted here.

Reply:
Right. I was thinking of turning the macro or function into a validation function (akin to the ones used in `rcl`, see [here](https://github.com/ros2/rcl/blob/master/rcl/include/rcl/client.h#L378)) and then use it accordingly i.e. to return early, to branch, etc. But it's a minor nit. I won't block on it.